### PR TITLE
SONAR-7876 Add an API for Sensors to report files that can't be analyzed

### PR DESCRIPTION
**Pull request by dbmeneses**

### CODE REVIEW

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/sensor/DefaultSensorContext.java [hunk 1]
```diff
@@ -31,6 +31,8 @@
 import org.sonar.api.batch.sensor.coverage.internal.DefaultCoverage;
 import org.sonar.api.batch.sensor.cpd.NewCpdTokens;
 import org.sonar.api.batch.sensor.cpd.internal.DefaultCpdTokens;
+import org.sonar.api.batch.sensor.error.NewAnalysisError;
+import org.sonar.api.batch.sensor.error.internal.DefaultAnalysisError;
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/sensor/DefaultSensorContext.java]: MINOR Remove this unused import 'org.sonar.api.batch.sensor.error.internal.DefaultAnalysisError'. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [error](https://github.com/SonarSource/sonarqube/blob/151bb85d038e511874cef019449eadc42917ef07/sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/error): Add a 'package-info.java' file to document the 'error' package [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1228)
1. MAJOR [internal](https://github.com/SonarSource/sonarqube/blob/151bb85d038e511874cef019449eadc42917ef07/sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/error/internal): Add a 'package-info.java' file to document the 'internal' package [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1228)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [error](https://github.com/SonarSource/sonarqube/blob/388c13eb5a5d00deeb7c1a9f34594bcc7d5ae372/sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/error): Add a 'package-info.java' file to document the 'error' package [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1228)
1. MAJOR [internal](https://github.com/SonarSource/sonarqube/blob/388c13eb5a5d00deeb7c1a9f34594bcc7d5ae372/sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/error/internal): Add a 'package-info.java' file to document the 'internal' package [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1228)

File: plugins/sonar-xoo-plugin/src/main/java/org/sonar/xoo/rule/AnalysisErrorSensor.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    try {
+      int line = Integer.parseInt(textPointer[0]);
+      int lineOffset = Integer.parseInt(textPointer[1]);
+      String msg = textPointer[2];
+
+      context.newAnalysisError()
+        .onFile(inputFile)
+        .at(new DefaultTextPointer(line, lineOffset))
```
Comment (henryju) [inline plugins/sonar-xoo-plugin/src/main/java/org/sonar/xoo/rule/AnalysisErrorSensor.java]: DefaultTextPointer is not an API. Use InputFile::newPointer

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/error/NewAnalysisError.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+   */
+  NewAnalysisError message(String message);
+
+  /**
+   * Location of this error. This field is optional.
+   */
+  NewAnalysisError at(TextPointer location);
+
+  /**
+   * Save this error information.
```
Comment (henryju) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/error/NewAnalysisError.java]: Maybe we could document that:
  * it is supported to save an analysis error several times for the same file, but behavior is not specified (ie do we display all errors, or only the first/last one)
  * it is supported to report issues and analysis error for the same file, but again result is not specified (do we display issues or not?)

**Comment by dbmeneses** [comment]
@henryju I've applied the feedback

**Comment by henryju** [comment]
LGTM but adding a medium test would be great!

**Pull request merged**